test/integration/grant-types/authorization-code-grant-type_test.js

```diff
@@ -361,6 +363,114 @@ describe('AuthorizationCodeGrantType integration', function() {
         });
     });
 
+    describe('with PKCE', function() {
+      it('should throw an error if the `code_verifier` is invalid with S256 code challenge method', function() {
+        var codeVerifier = stringUtil.base64URLEncode(crypto.randomBytes(32));
+        var authorizationCode = {
+          authorizationCode: 12345,
+          client: { id: 'foobar' },
+          expiresAt: new Date(new Date().getTime() * 2),
+          user: {},
+          codeChallengeMethod: 'S256',
+          codeChallenge: stringUtil.base64URLEncode(crypto.createHash('sha256').update(codeVerifier).digest())
+        };
+        var client = { id: 'foobar', isPublic: true };
+        var model = {
+          getAuthorizationCode: function() { return authorizationCode; },
+          revokeAuthorizationCode: function() {},
+          saveToken: function() {}
+        };
+        var grantType = new AuthorizationCodeGrantType({ accessTokenLifetime: 123, model: model });
+        var request = new Request({ body: { code: 12345, code_verifier: 'foo' }, headers: {}, method: {}, query: {} });
+
+        return grantType.getAuthorizationCode(request, client)
+          .then(should.fail)
+          .catch(function(e) {
+            e.should.be.an.instanceOf(InvalidGrantError);
+            e.message.should.equal('Invalid grant: code verifier is invalid');
+          });
+      });
+
+      it('should throw an error if the `code_verifier` is invalid with plain code challenge method', function() {
+        var codeVerifier = stringUtil.base64URLEncode(crypto.randomBytes(32));
+        var authorizationCode = {
+          authorizationCode: 12345,
+          client: { id: 'foobar' },
+          expiresAt: new Date(new Date().getTime() * 2),
+          user: {},
+          codeChallengeMethod: 'plain',
+          codeChallenge: codeVerifier
+        };
+        var client = { id: 'foobar', isPublic: true };
```

Comment:
The `isPublic` option is not used, as a result any client which allows `authorization_code` grant also accepts PKCE requests.